src/runtime/ndarray.cc

```diff
@@ -67,7 +67,8 @@ struct NDArray::Internal {
       ptr->mem = nullptr;
     } else if (ptr->dl_tensor.data != nullptr) {
       // if the array is still pinned before freeing, unpin it.
-      UnpinContainer(ptr);
+      if (ptr->pinned_by_dgl_)
```

Comment:
why not do the `if` check within `UnpinContainer()`? One more thing, always wrapping body of `if` statement is a better practice even there's single line...